cram/cram_io.c

```diff
@@ -4403,8 +4410,11 @@ int cram_set_voption(cram_fd *fd, enum hts_fmt_option opt, va_list args) {
 	break;
 
     case CRAM_OPT_RANGE:
+	pthread_mutex_lock(&fd->range_lock);
 	fd->range = *va_arg(args, cram_range *);
-	return cram_seek_to_refpos(fd, &fd->range);
+	pthread_mutex_unlock(&fd->range_lock);
+	int ret = cram_seek_to_refpos(fd, &fd->range);
```

Comment:
I think there could be a race here, where `cram_decode_slice()` in a thread tries to use the new `fd->range` before `cram_seek()` is able to drain the job from the queue.  I guess it might not matter as the results are thrown away, but it could be avoided by passing a `cram_range *` to somewhere other than `&fd->range` to `cram_seek_to_refpos`, and then getting that to fill in `fd->range` after it's done the seek.  Insisting that the second argument to `cram_seek_to_refpos()` is not `&fd->range` would allow some of the locks in that function to be removed as well.

Reply:
I don't think the race matters as we're immediately tearing down the threads anyway.  If they erroneously start reading a block that they didn't need because the range increased then it'll be short lived.  However I agree it can be simplified and will do so.